### PR TITLE
FIX More func base globals fixes

### PR DESCRIPTION
**Pull request by pierreglaser**
Follow up on #192, itself being a follow up on #187 

The goal is to keep track of global variable modifications in the child
environment where functions are unpickled, and to prevent those modifications to be overriden if an object containing those globals gets reloaded.  

NOTE: The case where the global variables live in `__main__` was handled in  [here](https://github.com/cloudpipe/cloudpickle/pull/188) by @tomMoral 
for the case where they live in any other module, a solution was proposed [here](https://github.com/cloudpipe/cloudpickle/pull/192)  by @ogrisel. The principle was to keep the name of the module where the globals lived instead of the variables themselves, and to call `vars(sys.modules[module_name])` at unpickling time only. 

However, the module where the pickled object lived in the parent environment does not always exist in the child environment. So `sys.modules` has to be checked in the new environment. 
If the module does not exist, `base_globals` start to get tracked in case of further
re-loadings, using `_BASE_GLOBALS_CACHE`

Ultimately, this was causing an issue in python 2 in the case of dynamic
modules, because `dynamic_subimport`, that creates the dynamic module, writes into sys.modules, which can cause recursive imports in python 2.

### CODE REVIEW

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -1090,7 +1098,16 @@ def _make_skel_func(code, cell_count, base_globals=None):
     if base_globals is None:
         base_globals = {}
     elif isinstance(base_globals, str):
-        base_globals = vars(sys.modules[base_globals])
+        if base_globals in sys.modules.keys():
+            # this checks if we can import the previous environment the object
+            # lived in
+            base_globals = vars(sys.modules[base_globals])
+        elif id(base_globals) in _BASE_GLOBALS_CACHE.keys():
+            base_globals = _BASE_GLOBALS_CACHE[id(base_globals)]
```
Comment (tomMoral) [inline cloudpickle/cloudpickle.py]: Could you test this mechanism?

Comment (pierreglaser) [reply cloudpickle/cloudpickle.py]: yes, was about to do it!

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -934,7 +939,10 @@ def subimport(name):
 def dynamic_subimport(name, vars):
     mod = imp.new_module(name)
     mod.__dict__.update(vars)
-    sys.modules[name] = mod
+    # adding a dynamic module to sys.module can cause recursive imports
+    # in python 2.x.x because site.Quitter contains sys as one of its globals
+    if sys.version > '3':
+        sys.modules[name] = mod
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: Mutating `sys.modules` only under Python 3 seems weird to me. Why the discrepancy between the two major versions? Why is this not required for Python 2?

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -78,6 +78,10 @@
     PY3 = True
 
 
+# cache that tracks the values of global variables.
+_BASE_GLOBALS_CACHE = {}
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: I am worried that this module level dict will cause memory leaks if loky is used in long-running processes with dynamic modules. Is this cache really necessary?

File: cloudpickle/cloudpickle.py [hunk 4]
```diff
@@ -1090,7 +1098,16 @@ def _make_skel_func(code, cell_count, base_globals=None):
     if base_globals is None:
         base_globals = {}
     elif isinstance(base_globals, str):
-        base_globals = vars(sys.modules[base_globals])
+        if base_globals in sys.modules.keys():
+            # this checks if we can import the previous environment the object
+            # lived in
+            base_globals = vars(sys.modules[base_globals])
+        elif id(base_globals) in _BASE_GLOBALS_CACHE.keys():
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: You should not use the id of an object that can be garbage collected as a key in a dict: this object id can be reused by Python for a subsequent object.

**Comment by pierreglaser** [comment]
@ogrisel I wrote a test that shows the use case of `_BASE_GLOBALS_CACHE`. 
In summary, when pickling a function living in a dynamic module, it will no longer be in sys.modules, so calling `vars(sys.modules[name])` at loading time won't work, and globals variables changes will be overriden.  Hence this cache creation, that is basically an alternative to sys.modules.

**Comment by ogrisel** [comment]
I think this should be renamed to `_dynamic_modules` and use the module name as the key rather than `id(module_name)` that could be recycled for another object in case of garbage collection.

**Comment by pierreglaser** [comment]
I switched from `id(module_name)` to `module_name` in the last commit. 
Noted for the _dynamic_modules comment.

File: cloudpickle/cloudpickle.py [hunk 5]
```diff
@@ -1090,7 +1094,17 @@ def _make_skel_func(code, cell_count, base_globals=None):
     if base_globals is None:
         base_globals = {}
     elif isinstance(base_globals, str):
-        base_globals = vars(sys.modules[base_globals])
+        base_globals_name = base_globals
+        if base_globals_name in sys.modules.keys():
+            # this checks if we can import the previous environment the object
+            # lived in
+            base_globals = vars(sys.modules[base_globals_name])
+        elif base_globals_name in _dynamic_modules.keys():
```
Comment (tomMoral) [inline cloudpickle/cloudpickle.py]: You could use the `get` function to simplify the code:

```python
base_globals = _dynamic_modules_globals.get(base_globals_name], {})
_dynamic_modules_globals[base_globals_name] = base_globals
```

I would also rename `_dynamic_modules` to `_dynamic_module_globals`

File: tests/cloudpickle_test.py [hunk 6]
```diff
@@ -887,6 +890,66 @@ def f1():
...
+        '''
+        exec(textwrap.dedent(code), mod.__dict__)
+
+        # first, dump the module
+        fileobj = cloudpickle.dumps(mod.func_that_relies_on_dynamic_module)
+
+        first_f = pickle.loads(fileobj)
+        # change the mod's global variable x
+        mod.x = 2
+
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: Maybe add an assert that `first_f() == 2`

File: tests/cloudpickle_test.py [hunk 7]
```diff
@@ -887,6 +890,66 @@ def f1():
...
+        fileobj = cloudpickle.dumps(mod.func_that_relies_on_dynamic_module)
+
+        mod.x = 1
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: I would put `mod.x = 3` and then test that `first_f() == 3`

File: tests/cloudpickle_test.py [hunk 8]
```diff
@@ -887,6 +890,66 @@ def f1():
...
+
+        # finally, re-load the dynamic module
+        new_f = pickle.loads(fileobj)
+
+        assert first_f() == new_f()
```
Comment (tomMoral) [inline tests/cloudpickle_test.py]: And also verify the value here as one issue can be that unpickling `new_f` might reset the global variable value to the initial value `1`.

**Comment by pierreglaser** [comment]
as discussed with Olivier and Tom, dynamic modules caching is adressed in [another PR](https://github.com/cloudpipe/cloudpickle/pull/201). 
Tests about dynamic modules global changes have been modified to comply with the standard behavior (e.g each time a function from a dynamic module is unpickled, a new set of `globals` gets created).

File: cloudpickle/cloudpickle.py [hunk 9]
```diff
@@ -78,6 +78,7 @@
     PY3 = True
 
 
+
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: Please do no insert an extra blank line. PEP8 only require 2 blank lines to separate top level blocks.

File: cloudpickle/cloudpickle.py [hunk 10]
```diff
@@ -934,7 +936,6 @@ def subimport(name):
 def dynamic_subimport(name, vars):
     mod = imp.new_module(name)
     mod.__dict__.update(vars)
-    sys.modules[name] = mod
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: This should not be changed in this PR, right? This line come from a very old commit (e7341b67).

The fact that this deleted line does not break any existing test is worrisome but this should not be changed by this PR in my opinion.

File: cloudpickle/cloudpickle.py (hunk 9 above)
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: Please do not add this extra blank line.

File: CHANGES.md [hunk 11]
```diff
@@ -1,7 +1,10 @@
-
 master
 ======
 
+- global variables from modules referenced in sys.modules in the child process
+  now overrides the initial global variables of the pickled function.
+  ([issue #187](https://github.com/cloudpipe/cloudpickle/issues/187))
```
Comment (ogrisel) [inline CHANGES.md]: I would rephrase this as:

"""
Ensure that unpickling a locally defined function that accesses the global variables of a module does not reset the values of the global variables if they are already initialized.
"""

File: tests/cloudpickle_test.py [hunk 12]
```diff
@@ -887,6 +891,107 @@ def f1():
                                         clone_func=clone_func)
             assert_run_python_script(textwrap.dedent(code))
 
+    def test_closure_interacting_with_a_global_variable(self):
+        global _TEST_GLOBAL_VARIABLE
+        orig_value = _TEST_GLOBAL_VARIABLE
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: maybe add:

```python
assert _TEST_GLOBAL_VARIABLE == "default_value"
```

to make the test both more easy to follow and robust to side effects by other tests.

File: tests/cloudpickle_test.py [hunk 13]
```diff
@@ -887,6 +891,107 @@ def f1():
...
+            _TEST_GLOBAL_VARIABLE = orig_value
+
+    def test_function_from_dynamic_module_with_globals_modifications(self):
+        """
+        unpickling functions coming from a dynamic module should lead to
+        new global variables being created each time.
+        Hence, a modification of the globals of one function should not
+        interact with the globals of another function.
+        This test validates this behavior.
+        """
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: Honestly I don't know what "should" happen in this case. I would rather remove this test from this PR otherwise it might give the impression that the current behavior is on purpose.

File: cloudpickle/cloudpickle.py (hunk 10 above)
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Actually, registering dynamic modules was a bug. Once registered as such, subsequent calls to dumps will consider the module as not dynamic which will make it impossible to load in other python processes.

File: tests/cloudpickle_test.py [hunk 14]
```diff
@@ -436,6 +440,66 @@ def method(self, x):
         mod1, mod2 = pickle_depickle([mod, mod])
         self.assertEqual(id(mod1), id(mod2))
 
+    def test_load_dynamic_module_in_grandchild_process(self):
+        # make sure that a when loaded, a dynamic module
+        # preserves its dynamic property. Otherwise,
+        # this will lead to an import error if pickled in the child process
+        # and reloaded in another one
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: Cosmetics: please do no insert new line too early in comments. Try to use your 80 char budgets. Also please start sentences with a capital case :)

File: tests/cloudpickle_test.py [hunk 15]
```diff
@@ -436,6 +440,66 @@ def method(self, x):
...
+            import textwrap
+
+            import cloudpickle
+            from testutils import assert_run_python_script
+
+
+            child_of_child_process_script = {}
+
+            with open('dynamic_module_from_parent_process.pk', 'rb') as fid:
+                mod = pickle.load(fid)
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: nitpick: fid => f. This is not an integer but a file object.

File: tests/cloudpickle_test.py [hunk 16]
```diff
@@ -436,6 +440,66 @@ def method(self, x):
...
+
+            with open('dynamic_module_from_child_process.pk', 'wb') as fid:
+                cloudpickle.dump(mod, fid)
+
+            assert_run_python_script(textwrap.dedent(child_of_child_process_script))
+            '''
+
+        # the script ran by the process created by the child process
+        child_of_child_process_script = ''' \'\'\'
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: Use triple double quote on the outside `"""` so that you can use nested triple single quotes `'''` in the inside.

File: tests/cloudpickle_test.py [hunk 17]
```diff
@@ -436,6 +440,66 @@ def method(self, x):
...
+            import pickle
+            import textwrap
+
+            import cloudpickle
+            from testutils import assert_run_python_script
+
+
+            child_of_child_process_script = {}
+
+            with open('dynamic_module_from_parent_process.pk', 'rb') as fid:
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: nitpick: use `.pkl` as filename extension for pickle files.

File: tests/cloudpickle_test.py [hunk 18]
```diff
@@ -436,6 +440,66 @@ def method(self, x):
...
+                cloudpickle.dump(mod, fid)
+
+            assert_run_python_script(textwrap.dedent(child_process_script))
+
+        finally:
+            # remove temporary created files
+            if os.path.exists('dynamic_module_from_parent_process.pk'):
+                os.unlink('dynamic_module_from_parent_process.pk')
+            if os.path.exists('dynamic_module_from_child_process.pk'):
+                os.unlink('dynamic_module_from_child_process.pk')
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: put those two filenames in local variables to avoid duplication and avoid hard to debug typos.

**Comment by mrocklin** [comment]
This may have caused a regression described here: https://github.com/cloudpipe/cloudpickle/issues/202
